File: api/library/comprehensive/predictions.py
# =============================================================================
# EVENT PREDICTION ENGINE
# =============================================================================
# Year-by-year scoring for major life events:
#   - Marriage timing
#   - Divorce risk
#   - Children arrival
#   - Career peaks/changes
# Uses natal chart, luck pillars, and annual pillar interactions.
# =============================================================================

from typing import List, Dict, Tuple, Optional
from ..core import STEMS, BRANCHES
from ..derived import (
    ELEMENT_CYCLES, get_ten_god, get_all_branch_qi,
    STEM_ORDER, BRANCH_ORDER
)
from .models import EventPrediction, ChartData, LuckPillarInfo
from .shen_sha import HONG_LUAN_LOOKUP, TIAN_XI_LOOKUP, TAO_HUA_LOOKUP, _xun_void_branches


# =============================================================================
# ANNUAL PILLAR CALCULATOR
# =============================================================================

def get_annual_pillar(year: int) -> Tuple[str, str]:
    """Calculate the Heavenly Stem and Earthly Branch for a given year."""
    stem_idx = (year - 4) % 10
    branch_idx = (year - 4) % 12
    return STEM_ORDER[stem_idx], BRANCH_ORDER[branch_idx]


# =============================================================================
# CLASH DETECTION HELPERS
# =============================================================================

CLASH_MAP = {b: BRANCHES[b]["clashes"] for b in BRANCHES}
HARMONY_MAP = {b: BRANCHES[b]["harmonizes"] for b in BRANCHES}


def _branches_clash(b1: str, b2: str) -> bool:
    return CLASH_MAP.get(b1) == b2


def _branches_harmonize(b1: str, b2: str) -> bool:
    return HARMONY_MAP.get(b1) == b2


# =============================================================================
# MARRIAGE TIMING PREDICTOR
# =============================================================================

def predict_marriage_years(chart: ChartData, start_year: int = 0,
                           end_year: int = 0) -> List[EventPrediction]:
    """
    Score each year for marriage probability.
    Factors:
    1. Hong Luan (红鸾) or Tian Xi (天喜) activated by annual branch
    2. Peach Blossom activated by annual branch
    3. Spouse Palace (Day Branch) harmonized or clashed by annual branch
    4. Spouse star (DW for male, DO for female) appearing in annual stem
    5. Luck pillar alignment
    6. Cultural age window bonus
    """
    if start_year == 0:
        start_year = chart.birth_year + 18
    if end_year == 0:
        end_year = chart.birth_year + 55

    dm = chart.day_master
    day_branch = chart.pillars["day"].branch
    year_branch = chart.pillars["year"].branch

    # Spouse star for this gender
    spouse_star_abbr = "DW" if chart.gender == "male" else "DO"

    # Hong Luan and Tian Xi targets
    hl_target = HONG_LUAN_LOOKUP.get(year_branch)
    tx_target = TIAN_XI_LOOKUP.get(year_branch)

    # Peach Blossom targets
    pb_targets = set()
    for base in ["year", "day"]:
        base_br = chart.pillars[base].branch
        pb = TAO_HUA_LOOKUP.get(base_br)
        if pb:
            pb_targets.add(pb)

    results = []

    for year in range(start_year, end_year + 1):
        age = year - chart.birth_year
        score = 0.0
        factors = []

        annual_stem, annual_branch = get_annual_pillar(year)

        # 1. Hong Luan activated
        if annual_branch == hl_target:
            score += 25
            factors.append("Hong Luan (红鸾) activated")

        # 2. Tian Xi activated
        if annual_branch == tx_target:
            score += 20
            factors.append("Tian Xi (天喜) activated")

        # 3. Peach Blossom activated
        if annual_branch in pb_targets:
            score += 15
            factors.append("Peach Blossom (桃花) activated")

        # 4. Spouse palace (day branch) harmonized
        if _branches_harmonize(day_branch, annual_branch):
            score += 20
            factors.append("Spouse palace harmonized by annual branch")

        # 5. Spouse palace clashed (can trigger marriage OR separation)
        if _branches_clash(day_branch, annual_branch):
            score += 10
            factors.append("Spouse palace clashed (can trigger marriage event)")

        # 6. Spouse star in annual stem
        tg = get_ten_god(dm, annual_stem)
        if tg and tg[0] == spouse_star_abbr:
            score += 20
            factors.append(f"Spouse star ({spouse_star_abbr}) in annual stem")

        # 7. Spouse star in annual branch hidden stems
        for hs, qs in get_all_branch_qi(annual_branch):
            tg = get_ten_god(dm, hs)
            if tg and tg[0] == spouse_star_abbr:
                score += 10
                factors.append(f"Spouse star in annual branch hidden stem")
                break

        # 8. Cultural age window
        if chart.gender == "male":
            if 26 <= age <= 33:
                score += 10
                factors.append("Prime marriage age window (male 26-33)")
            elif 22 <= age <= 25:
                score += 5
        else:
            if 23 <= age <= 29:
                score += 10
                factors.append("Prime marriage age window (female 23-29)")
            elif 20 <= age <= 22:
                score += 5

        # 9. Check luck pillar alignment
        for lp in chart.luck_pillars:
            if lp.start_year <= year <= lp.end_year:
                lp_tg = get_ten_god(dm, lp.stem)
                if lp_tg and lp_tg[0] == spouse_star_abbr:
                    score += 10
                    factors.append(f"Luck pillar stem = spouse star")
                if _branches_harmonize(day_branch, lp.branch):
                    score += 5
                    factors.append(f"Luck pillar harmonizes spouse palace")
                break

        if score >= 20:
            results.append(EventPrediction(
                event_type="marriage",
                year=year,
                age=age,
                score=round(score, 1),
                factors=factors,
            ))

    results.sort(key=lambda x: x.score, reverse=True)
    return results


# =============================================================================
# DIVORCE RISK PREDICTOR
# =============================================================================

def predict_divorce_years(chart: ChartData, start_year: int = 0,
                          end_year: int = 0) -> List[EventPrediction]:
    """
    Score each year for divorce/separation risk.
    Factors:
    1. Spouse palace clashed by annual branch
    2. Rob Wealth in annual stem (steals spouse)
    3. Hurting Officer in annual stem (female: destroys husband star)
    4. Punishment involving spouse palace
    5. Void on spouse palace activated
    """
    if start_year == 0:
        start_year = chart.birth_year + 20
    if end_year == 0:
        end_year = chart.birth_year + 60

    dm = chart.day_master
    day_branch = chart.pillars["day"].branch

    results = []

    for year in range(start_year, end_year + 1):
        age = year - chart.birth_year
        score = 0.0
        factors = []

        annual_stem, annual_branch = get_annual_pillar(year)

        # 1. Spouse palace clashed
        if _branches_clash(day_branch, annual_branch):
            score += 25
            factors.append("Spouse palace directly clashed")

        # 2. Rob Wealth in annual stem
        tg = get_ten_god(dm, annual_stem)
        if tg and tg[0] == "RW":
            score += 20
            factors.append("Rob Wealth appears (competitor for spouse)")

        # 3. Hurting Officer (especially bad for females)
        if tg and tg[0] == "HO":
            bonus = 20 if chart.gender == "female" else 10
            score += bonus
            factors.append("Hurting Officer appears (damages marriage structure)")

        # 4. Seven Killings for females (rival male)
        if chart.gender == "female" and tg and tg[0] == "7K":
            score += 10
            factors.append("Seven Killings appears (rival/affair indicator)")

        # 5. Check punishment patterns with annual branch
        natal_branches = [chart.pillars[p].branch for p in ["year", "month", "day", "hour"]]
        punishment_check = set(natal_branches) | {annual_branch}

        # Ungrateful punishment (Yin-Si-Shen)
        ungrateful = {"Yin", "Si", "Shen"}
        if len(ungrateful & punishment_check) >= 3 and annual_branch in ungrateful:
            score += 15
            factors.append("Ungrateful punishment activated")

        # Bullying punishment (Chou-Wei-Xu)
        bullying = {"Chou", "Wei", "Xu"}
        if len(bullying & punishment_check) >= 3 and annual_branch in bullying:
            score += 15
            factors.append("Bullying punishment activated")

        # 6. Luck pillar clash on spouse palace
        for lp in chart.luck_pillars:
            if lp.start_year <= year <= lp.end_year:
                if _branches_clash(day_branch, lp.branch):
                    score += 15
                    factors.append("Luck pillar also clashes spouse palace")
                break

        if score >= 20:
            results.append(EventPrediction(
                event_type="divorce",
                year=year,
                age=age,
                score=round(score, 1),
                factors=factors,
            ))

    results.sort(key=lambda x: x.score, reverse=True)
    return results


# =============================================================================
# CHILDREN ARRIVAL PREDICTOR
# =============================================================================

def predict_children_years(chart: ChartData, start_year: int = 0,
                           end_year: int = 0) -> List[EventPrediction]:
    """
    Score each year for children arrival probability.
    Male: 7K = sons, DO = daughters
    Female: HO = sons, EG = daughters
    """
    if start_year == 0:
        start_year = chart.birth_year + 20
    if end_year == 0:
        end_year = chart.birth_year + 50

    dm = chart.day_master
    hour_branch = chart.pillars["hour"].branch

    # Children star mapping
    if chart.gender == "male":
        child_stars = {"7K", "DO"}
    else:
        child_stars = {"HO", "EG"}

    results = []

    for year in range(start_year, end_year + 1):
        age = year - chart.birth_year
        score = 0.0
        factors = []

        annual_stem, annual_branch = get_annual_pillar(year)

        # 1. Children star in annual stem
        tg = get_ten_god(dm, annual_stem)
        if tg and tg[0] in child_stars:
            score += 25
            factors.append(f"Children star ({tg[0]}) in annual stem")

        # 2. Children star in annual branch hidden stems
        for hs, qs in get_all_branch_qi(annual_branch):
            tg_h = get_ten_god(dm, hs)
            if tg_h and tg_h[0] in child_stars:
                score += 10
                factors.append(f"Children star in annual branch")
                break

        # 3. Hour palace (children palace) harmonized
        if _branches_harmonize(hour_branch, annual_branch):
            score += 15
            factors.append("Children palace harmonized")

        # 4. Cultural age window
        if chart.gender == "male":
            if 27 <= age <= 36:
                score += 8
        else:
            if 24 <= age <= 33:
                score += 8

        # 5. Luck pillar alignment
        for lp in chart.luck_pillars:
            if lp.start_year <= year <= lp.end_year:
                lp_tg = get_ten_god(dm, lp.stem)
                if lp_tg and lp_tg[0] in child_stars:
                    score += 10
                    factors.append("Luck pillar stem = children star")
                break

        if score >= 20:
            results.append(EventPrediction(
                event_type="child_birth",
                year=year,
                age=age,
                score=round(score, 1),
                factors=factors,
            ))

    results.sort(key=lambda x: x.score, reverse=True)
    return results


# =============================================================================
# CAREER PEAK PREDICTOR
# =============================================================================

def predict_career_years(chart: ChartData, start_year: int = 0,
                         end_year: int = 0) -> List[EventPrediction]:
    """
    Score each year for career advancement probability.
    """
    if start_year == 0:
        start_year = chart.birth_year + 20
    if end_year == 0:
        end_year = chart.birth_year + 65

    dm = chart.day_master

    results = []

    for year in range(start_year, end_year + 1):
        age = year - chart.birth_year
        score = 0.0
        factors = []

        annual_stem, annual_branch = get_annual_pillar(year)
        tg = get_ten_god(dm, annual_stem)

        # 1. Direct Officer in annual stem (promotion, recognition)
        if tg and tg[0] == "DO":
            score += 20
            factors.append("Direct Officer year (recognition, promotion)")

        # 2. Seven Killings in annual stem (power, but risky)
        if tg and tg[0] == "7K":
            score += 15
            factors.append("Seven Killings year (power grab, risk)")

        # 3. Indirect Wealth in annual stem (windfall, business)
        if tg and tg[0] == "IW":
            score += 15
            factors.append("Indirect Wealth year (windfall, opportunity)")

        # 4. Month pillar (career palace) harmonized
        month_branch = chart.pillars["month"].branch
        if _branches_harmonize(month_branch, annual_branch):
            score += 15
            factors.append("Career palace harmonized")

        # 5. Career palace clashed (career upheaval - can be promotion or loss)
        if _branches_clash(month_branch, annual_branch):
            score += 10
            factors.append("Career palace clashed (career change event)")

        # 6. Luck pillar alignment
        for lp in chart.luck_pillars:
            if lp.start_year <= year <= lp.end_year:
                lp_tg = get_ten_god(dm, lp.stem)
                if lp_tg and lp_tg[0] in ("DO", "7K", "IW"):
                    score += 10
                    factors.append(f"Luck pillar = {lp_tg[0]} (career energy)")
                break

        if score >= 20:
            results.append(EventPrediction(
                event_type="career_peak",
                year=year,
                age=age,
                score=round(score, 1),
                factors=factors,
            ))

    results.sort(key=lambda x: x.score, reverse=True)
    return results


# =============================================================================
# MASTER PREDICTION FUNCTION
# =============================================================================

def run_all_predictions(chart: ChartData) -> Dict[str, List[EventPrediction]]:
    """Run all prediction algorithms and return top results for each category."""
    return {
        "marriage": predict_marriage_years(chart)[:5],
        "divorce": predict_divorce_years(chart)[:5],
        "children": predict_children_years(chart)[:5],
        "career": predict_career_years(chart)[:5],
    }

File: src/lib/bazi/comprehensive/predictions.ts

// =============================================================================
// EVENT PREDICTION ENGINE
// =============================================================================
// Year-by-year scoring for major life events:
//   - Marriage timing
//   - Divorce risk
//   - Children arrival
//   - Career peaks/changes
// Uses natal chart, luck pillars, and annual pillar interactions.
// =============================================================================

import { BRANCHES, type BranchName, type StemName } from '../core';
import { getTenGod, getAllBranchQi, STEM_ORDER, BRANCH_ORDER } from '../derived';
import type { EventPrediction, ChartData } from './models';
import { HONG_LUAN_LOOKUP, TIAN_XI_LOOKUP, TAO_HUA_LOOKUP } from './shen-sha';

// =============================================================================
// ANNUAL PILLAR CALCULATOR
// =============================================================================

export function getAnnualPillar(year: number): [StemName, BranchName] {
  /** Calculate the Heavenly Stem and Earthly Branch for a given year. */
  const stemIdx = ((year - 4) % 10 + 10) % 10;
  const branchIdx = ((year - 4) % 12 + 12) % 12;
  return [STEM_ORDER[stemIdx], BRANCH_ORDER[branchIdx]];
}

// =============================================================================
// CLASH DETECTION HELPERS
// =============================================================================

const CLASH_MAP: Record<string, string> = {};
const HARMONY_MAP: Record<string, string> = {};
for (const b of Object.keys(BRANCHES) as BranchName[]) {
  CLASH_MAP[b] = BRANCHES[b].clashes;
  HARMONY_MAP[b] = BRANCHES[b].harmonizes;
}

function _branchesClash(b1: string, b2: string): boolean {
  return CLASH_MAP[b1] === b2;
}

function _branchesHarmonize(b1: string, b2: string): boolean {
  return HARMONY_MAP[b1] === b2;
}

// =============================================================================
// MARRIAGE TIMING PREDICTOR
// =============================================================================

export function predictMarriageYears(
  chart: ChartData,
  startYear = 0,
  endYear = 0,
): EventPrediction[] {
  if (startYear === 0) startYear = chart.birth_year + 18;
  if (endYear === 0) endYear = chart.birth_year + 55;

  const dm = chart.day_master;
  const dayBranch = chart.pillars["day"].branch;
  const yearBranch = chart.pillars["year"].branch;

  const spouseStarAbbr = chart.gender === "male" ? "DW" : "DO";

  const hlTarget = HONG_LUAN_LOOKUP[yearBranch] ?? null;
  const txTarget = TIAN_XI_LOOKUP[yearBranch] ?? null;

  const pbTargets = new Set<string>();
  for (const base of ["year", "day"]) {
    const baseBr = chart.pillars[base].branch;
    const pb = TAO_HUA_LOOKUP[baseBr];
    if (pb) pbTargets.add(pb);
  }

  const results: EventPrediction[] = [];

  for (let year = startYear; year <= endYear; year++) {
    const age = year - chart.birth_year;
    let score = 0;
    const factors: string[] = [];

    const [annualStem, annualBranch] = getAnnualPillar(year);

    // 1. Hong Luan activated
    if (annualBranch === hlTarget) {
      score += 25;
      factors.push("Hong Luan (红鸾) activated");
    }

    // 2. Tian Xi activated
    if (annualBranch === txTarget) {
      score += 20;
      factors.push("Tian Xi (天喜) activated");
    }

    // 3. Peach Blossom activated
    if (pbTargets.has(annualBranch)) {
      score += 15;
      factors.push("Peach Blossom (桃花) activated");
    }

    // 4. Spouse palace harmonized
    if (_branchesHarmonize(dayBranch, annualBranch)) {
      score += 20;
      factors.push("Spouse palace harmonized by annual branch");
    }

    // 5. Spouse palace clashed
    if (_branchesClash(dayBranch, annualBranch)) {
      score += 10;
      factors.push("Spouse palace clashed (can trigger marriage event)");
    }

    // 6. Spouse star in annual stem
    const tg = getTenGod(dm, annualStem);
    if (tg && tg[0] === spouseStarAbbr) {
      score += 20;
      factors.push(`Spouse star (${spouseStarAbbr}) in annual stem`);
    }

    // 7. Spouse star in annual branch hidden stems
    for (const [hs] of getAllBranchQi(annualBranch)) {
      const tgH = getTenGod(dm, hs);
      if (tgH && tgH[0] === spouseStarAbbr) {
        score += 10;
        factors.push("Spouse star in annual branch hidden stem");
        break;
      }
    }

    // 8. Cultural age window
    if (chart.gender === "male") {
      if (age >= 26 && age <= 33) {
        score += 10;
        factors.push("Prime marriage age window (male 26-33)");
      } else if (age >= 22 && age <= 25) {
        score += 5;
      }
    } else {
      if (age >= 23 && age <= 29) {
        score += 10;
        factors.push("Prime marriage age window (female 23-29)");
      } else if (age >= 20 && age <= 22) {
        score += 5;
      }
    }

    // 9. Luck pillar alignment
    for (const lp of chart.luck_pillars) {
      if (lp.start_year <= year && year <= lp.end_year) {
        const lpTg = getTenGod(dm, lp.stem);
        if (lpTg && lpTg[0] === spouseStarAbbr) {
          score += 10;
          factors.push("Luck pillar stem = spouse star");
        }
        if (_branchesHarmonize(dayBranch, lp.branch)) {
          score += 5;
          factors.push("Luck pillar harmonizes spouse palace");
        }
        break;
      }
    }

    if (score >= 20) {
      results.push({
        event_type: "marriage",
        year,
        age,
        score: Math.round(score * 10) / 10,
        factors,
      });
    }
  }

  results.sort((a, b) => b.score - a.score);
  return results;
}

// =============================================================================
// DIVORCE RISK PREDICTOR
// =============================================================================

export function predictDivorceYears(
  chart: ChartData,
  startYear = 0,
  endYear = 0,
): EventPrediction[] {
  if (startYear === 0) startYear = chart.birth_year + 20;
  if (endYear === 0) endYear = chart.birth_year + 60;

  const dm = chart.day_master;
  const dayBranch = chart.pillars["day"].branch;
  const results: EventPrediction[] = [];

  for (let year = startYear; year <= endYear; year++) {
    const age = year - chart.birth_year;
    let score = 0;
    const factors: string[] = [];

    const [annualStem, annualBranch] = getAnnualPillar(year);

    // 1. Spouse palace clashed
    if (_branchesClash(dayBranch, annualBranch)) {
      score += 25;
      factors.push("Spouse palace directly clashed");
    }

    // 2. Rob Wealth in annual stem
    const tg = getTenGod(dm, annualStem);
    if (tg && tg[0] === "RW") {
      score += 20;
      factors.push("Rob Wealth appears (competitor for spouse)");
    }

    // 3. Hurting Officer
    if (tg && tg[0] === "HO") {
      const bonus = chart.gender === "female" ? 20 : 10;
      score += bonus;
      factors.push("Hurting Officer appears (damages marriage structure)");
    }

    // 4. Seven Killings for females
    if (chart.gender === "female" && tg && tg[0] === "7K") {
      score += 10;
      factors.push("Seven Killings appears (rival/affair indicator)");
    }

    // 5. Punishment patterns
    const natalBranches = ["year", "month", "day", "hour"].map(p => chart.pillars[p].branch);
    const punishmentCheck = new Set([...natalBranches, annualBranch]);

    const ungrateful = new Set(["Yin", "Si", "Shen"]);
    const ungratefulIntersect = [...ungrateful].filter(x => punishmentCheck.has(x as BranchName));
    if (ungratefulIntersect.length >= 3 && ungrateful.has(annualBranch)) {
      score += 15;
      factors.push("Ungrateful punishment activated");
    }

    const bullying = new Set(["Chou", "Wei", "Xu"]);
    const bullyingIntersect = [...bullying].filter(x => punishmentCheck.has(x as BranchName));
    if (bullyingIntersect.length >= 3 && bullying.has(annualBranch)) {
      score += 15;
      factors.push("Bullying punishment activated");
    }

    // 6. Luck pillar clash
    for (const lp of chart.luck_pillars) {
      if (lp.start_year <= year && year <= lp.end_year) {
        if (_branchesClash(dayBranch, lp.branch)) {
          score += 15;
          factors.push("Luck pillar also clashes spouse palace");
        }
        break;
      }
    }

    if (score >= 20) {
      results.push({
        event_type: "divorce",
        year,
        age,
        score: Math.round(score * 10) / 10,
        factors,
      });
    }
  }

  results.sort((a, b) => b.score - a.score);
  return results;
}

// =============================================================================
// CHILDREN ARRIVAL PREDICTOR
// =============================================================================

export function predictChildrenYears(
  chart: ChartData,
  startYear = 0,
  endYear = 0,
): EventPrediction[] {
  if (startYear === 0) startYear = chart.birth_year + 20;
  if (endYear === 0) endYear = chart.birth_year + 50;

  const dm = chart.day_master;
  const hourBranch = chart.pillars["hour"].branch;

  const childStars = new Set(
    chart.gender === "male" ? ["7K", "DO"] : ["HO", "EG"]);

  const results: EventPrediction[] = [];

  for (let year = startYear; year <= endYear; year++) {
    const age = year - chart.birth_year;
    let score = 0;
    const factors: string[] = [];

    const [annualStem, annualBranch] = getAnnualPillar(year);

    // 1. Children star in annual stem
    const tg = getTenGod(dm, annualStem);
    if (tg && childStars.has(tg[0])) {
      score += 25;
      factors.push(`Children star (${tg[0]}) in annual stem`);
    }

    // 2. Children star in annual branch hidden stems
    for (const [hs] of getAllBranchQi(annualBranch)) {
      const tgH = getTenGod(dm, hs);
      if (tgH && childStars.has(tgH[0])) {
        score += 10;
        factors.push("Children star in annual branch");
        break;
      }
    }

    // 3. Hour palace harmonized
    if (_branchesHarmonize(hourBranch, annualBranch)) {
      score += 15;
      factors.push("Children palace harmonized");
    }

    // 4. Cultural age window
    if (chart.gender === "male") {
      if (age >= 27 && age <= 36) score += 8;
    } else {
      if (age >= 24 && age <= 33) score += 8;
    }

    // 5. Luck pillar alignment
    for (const lp of chart.luck_pillars) {
      if (lp.start_year <= year && year <= lp.end_year) {
        const lpTg = getTenGod(dm, lp.stem);
        if (lpTg && childStars.has(lpTg[0])) {
          score += 10;
          factors.push("Luck pillar stem = children star");
        }
        break;
      }
    }

    if (score >= 20) {
      results.push({
        event_type: "child_birth",
        year,
        age,
        score: Math.round(score * 10) / 10,
        factors,
      });
    }
  }

  results.sort((a, b) => b.score - a.score);
  return results;
}

// =============================================================================
// CAREER PEAK PREDICTOR
// =============================================================================

export function predictCareerYears(
  chart: ChartData,
  startYear = 0,
  endYear = 0,
): EventPrediction[] {
  if (startYear === 0) startYear = chart.birth_year + 20;
  if (endYear === 0) endYear = chart.birth_year + 65;

  const dm = chart.day_master;
  const results: EventPrediction[] = [];

  for (let year = startYear; year <= endYear; year++) {
    const age = year - chart.birth_year;
    let score = 0;
    const factors: string[] = [];

    const [annualStem, annualBranch] = getAnnualPillar(year);
    const tg = getTenGod(dm, annualStem);

    // 1. Direct Officer
    if (tg && tg[0] === "DO") {
      score += 20;
      factors.push("Direct Officer year (recognition, promotion)");
    }

    // 2. Seven Killings
    if (tg && tg[0] === "7K") {
      score += 15;
      factors.push("Seven Killings year (power grab, risk)");
    }

    // 3. Indirect Wealth
    if (tg && tg[0] === "IW") {
      score += 15;
      factors.push("Indirect Wealth year (windfall, opportunity)");
    }

    // 4. Month pillar (career palace) harmonized
    const monthBranch = chart.pillars["month"].branch;
    if (_branchesHarmonize(monthBranch, annualBranch)) {
      score += 15;
      factors.push("Career palace harmonized");
    }

    // 5. Career palace clashed
    if (_branchesClash(monthBranch, annualBranch)) {
      score += 10;
      factors.push("Career palace clashed (career change event)");
    }

    // 6. Luck pillar alignment
    for (const lp of chart.luck_pillars) {
      if (lp.start_year <= year && year <= lp.end_year) {
        const lpTg = getTenGod(dm, lp.stem);
        if (lpTg && (lpTg[0] === "DO" || lpTg[0] === "7K" || lpTg[0] === "IW")) {
          score += 10;
          factors.push(`Luck pillar = ${lpTg[0]} (career energy)`);
        }
        break;
      }
    }

    if (score >= 20) {
      results.push({
        event_type: "career_peak",
        year,
        age,
        score: Math.round(score * 10) / 10,
        factors,
      });
    }
  }

  results.sort((a, b) => b.score - a.score);
  return results;
}

// =============================================================================
// MASTER PREDICTION FUNCTION
// =============================================================================

export function runAllPredictions(chart: ChartData): Record<string, EventPrediction[]> {
  /** Run all prediction algorithms and return top results for each category. */
  return {
    marriage: predictMarriageYears(chart).slice(0, 5),
    divorce: predictDivorceYears(chart).slice(0, 5),
    children: predictChildrenYears(chart).slice(0, 5),
    career: predictCareerYears(chart).slice(0, 5),
  };
}
